### DISCUSSION TITLE
Callback function as class member

### DISCUSSION THREAD

**Discussion by ghost** [Makuna/NeoPixelBus#576] [Q&A] [answered]
I'm trying to change the [FadeInOut](https://github.com/Makuna/NeoPixelBus/blob/master/examples/animations/NeoPixelFunFadeInOut/NeoPixelFunFadeInOut.ino) example.

The `BlendAnimUpdate` is a member function of my class:

```
void MyClass::BlendAnimUpdate(const AnimationParam& param)
{
    RgbColor updatedColor = RgbColor::LinearBlend(_animationState[param.index].StartingColor, _animationState[param.index].EndingColor, param.progress);
    _strip->SetPixelColor(_pos, updatedColor);
}
```

But this does not compile:

```
void MyClass::setColor(const char *name)
{
    HtmlColor color;
    color.Parse<HtmlShortColorNames>(name);

    _animationState[0].StartingColor = _strip->GetPixelColor(_pos);
    _animationState[0].EndingColor = color;
    _animation.StartAnimation(0, 500, BlendAnimUpdate);
}
```

> error: invalid use of non-static member function 'void MyClass::BlendAnimUpdate(const AnimationParam&)'

I really don't want to make my function static... is there a way to overcome this issue or is a limitation of the library?

**Comment by ghost** [comment]
Solved using a lambda:

```
    _animation.StartAnimation(0, 500, [this](const AnimationParam& param) {
        RgbColor updatedColor = RgbColor::LinearBlend(_animationState[param.index].StartingColor, _animationState[param.index].EndingColor, param.progress);
        _strip->SetPixelColor(_pos, updatedColor);
    });
```

Perhaps it can be documented somewhere. In real applications the features are organized in classes rather than putting everything outside.

**Comment by Makuna** [comment, marked as answer]
@Mark-81 (moved to Discussions where it should have been asked first)

The issue you ran into is a C++ language issue; it has very poor handling of callbacks and `this` capturing with function pointers.  Later, lambdas were added to help.  While your solution will work, a better one is listed below.

**Some Background:**
This link discusses the C++ callback issue with member functions:
https://stackoverflow.com/questions/14189440/c-callback-using-class-member

I can't rely on the functional header to be present in Arduino which gives you `std::function` and `std::bind`.  I found that many Arduino platforms didn't include it and the common std headers just didn't work with those platforms (probably why they didn't include it).  And since this library is cross platform, I had to omit the normal approach of using std.

**The best solution:**
My suggested model to use is to use the lambda `this` capturing ability to call the member.   It seems more concise of any of the solutions.  The code snippet follows on how to make the call. 

```
_animation.StartAnimation( 0, 500, [this](const AnimationParam& param) { BlendAnimUpdate( param); } );
```
C++ for the lambda captures the this pointer, and due this line of code (from your example above) is within the context of this class, it will call the member BlendAnimUpdate.
